docs/specifications/tests/DNSSEC-TP/dnssec16.md

```diff
@@ -0,0 +1,220 @@
+# DNSSEC16: Validate CDS
+
+## Test case identifier
+**DNSSEC16**
+
+## Objective
+
+CDS and CDNSKEY record types are defined in [RFC 7344] and [RFC 8078].
+Both record types are optional in a zone. The objective of this test
+case is to verify that the CDS RRset is valid. This test case is
+only relevant if the zone has at least on CDS record. For tests of the
+CDNSKEY, see test case [DNSSEC17].
+
+## Scope
+
+It is assumed that *Child Zone* has been tested by [Basic04]. This test
+case will just ignore non-responsive name servers or name servers not
+giving a correct DNS response for an authoritative name server.
+
+It is assumed that *Child Zone* has been tested or will be tested by 
+[DNSSEC15] and [DNSSEC17] and that the servers give the same responses.
+Running this test case without running [DNSSEC15] and [DNSSEC17] can
+give an incomplete report of the CDS and CDNSKEY status of 
+*Child Zone*.
+
+## Inputs
+
+* "Child Zone" - The domain name to be tested.
+
+## Summary
+
+* If no CDS record is found, the test case will terminate early
+  with no message tag outputted.
+* If a CDS record is of "delete" type, then it can by definition not
+  match or point at any DNSKEY record.
+
+Message Tag outputted                | [Default level] | Description of when message tag is outputted
+:------------------------------------|:--------|:-----------------------------------------
+DS16_CDS_UNSIGNED                    | ERROR   | CDS RRset is not signed.
+DS16_CDS_WITHOUT_DNSKEY              | ERROR   | CDS RRset exists, but no DNSKEY RRset.
+DS16_CDS_INVALID_RRSIG               | ERROR   | CDS RRset is signed with an invalid RRSIG.
+DS16_CDS_MATCHES_NO_DNSKEY           | WARNING | CDS record does not match any DNSKEY in DNSKEY RRset.
+DS16_CDS_SIGNED_BY_UNKNOWN_DNSKEY    | ERROR   | CDS RRset is signed but not by a key in DNSKEY RRset.
+DS16_DELETE_CDS                      | INFO    | CDS RRset has a "delete" CDS record as a single record.
+DS16_DNSKEY_NOT_SIGNED_BY_CDS        | WARNING | DNSKEY RRset is not signed by the key or keys that the CDS records point to.
+DS16_MIXED_DELETE_CDS                | ERROR   | "Delete" CDS record is mixed with normal CDS record.
+
+## Ordered description of steps to be taken to execute the test case
+
+1.  Create a CDS query with EDNS enabled and the DO bit set for the
+    apex of the *Child Zone*.
+
+2.  Create a DNSKEY query with EDNS enabled and the DO bit set for
+    the apex of the *Child Zone*.
+
+3.  Retrieve all name server IP addresses for the *Child Zone* using
+    [Method4] and [Method5] ("NS IP").
+
+4.  Create the following empty sets:
+    1.  Name server IP address and associated CDS RRset and its RRSIG
+        records ("CDS RRsets"). RRSIG records may be absent.
```

Comment:
Yes. My suggestion is to make it so that the RRSIG RRset is required, but that it may be empty. That way you only have one kind of absence to think about when you consider the state space that can be represented within the defined data structures. Having a smaller state space makes to consider makes the steps below easier to understand.

As authors of the test specifications we may know that the steps don't make use of the possibility for the RRSIG RRset to be present but empty. But for someone working through the test case who doesn't understand it yet, this distinction is another thing to keep in mind. And AFAICT for on good reason.

Reply:
Oh, that's interesting! You're thinking of an RRset as something different than a set of records of the same record type? Is there a meaningful distinction?

In any case the empty set is a very simple yet powerful idea that makes many things easier.